File: 0494-target-sum/0494-target-sum.cpp
class Solution {
private:
    int f(vector<int>& nums, int k, int i) {
        if (i == nums.size() - 1) {
            if (k == 0 && nums[i] == 0) {
                return 2;
            }
            if (k == 0 || nums[i] == k) {
                return 1;
            }
            return 0;
        }
        int nt = f(nums, k, i + 1);
        int t = 0;
        if (k >= nums[i]) {
            t = f(nums, k - nums[i], i + 1);
        }
        return t + nt;
    }

public:
    int findTargetSumWays(vector<int>& nums, int target) {
        int n = nums.size();
        int k = -1 * target;
        for (int i = 0; i < n; i++) {
            k += nums[i];
        }
        if (k < 0 || k & 1) {
            return 0;
        }
        k = k / 2;
        return f(nums, k, 0);
    }
};

File: 0494-target-sum/0494-target-sum.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        dp = [[-1] * 2001 for _ in range(len(nums))]
        def dfs(idx, summ):
            if idx == len(nums):
                if summ == target:
                    return 1
                else:
                    return 0
            if dp[idx][summ + 1000] != -1:
                return dp[idx][summ + 1000]
            
            dp[idx][summ + 1000] = dfs(idx + 1, summ + nums[idx]) + dfs(idx + 1, summ - nums[idx])
            return dp[idx][summ + 1000]
        
        return dfs(0, 0)